### ISSUE TITLE
Integrate openaddresses.io for search

### ISSUE THREAD

**Issue by zyphlar** [organicmaps/organicmaps#1263]
A major failing of nearly all OSM-based apps (besides maybe the OSM website itself with its supercharged search) is that only addresses entered in the database are searchable.

It may not be yet pragmatic to import all of https://openaddresses.io/ into OSM, but it should be relatively straightforward for client apps to use it as a data source. In fact, importing that data into OsmAnd is the only way I've been able to make OsmAnd usable for navigation.

If Organic Maps was able to use that data source to fill in the gaps of the OSM database during search, it would greatly improve the usability of the app and nearly eliminate the perrenial "can't find ____ address" issues (especially if common abbreviation matching was implemented, like N/North, Ave/Avenue, etc)

**Comment by biodranik** [comment]
What users will do if they see an empty map on the found address? Will they trust that search result?

**Comment by zyphlar** [comment]
You mean in case someone searches for an area they don't have downloaded? Openaddresses.io allows download of specific regions so it's easy to only include addresses that are already downloaded in Organic Maps. Or, if online searching and downloading is supported, that could be done automatically. Finally, even in a naive implementation, the map could simply say "this area isn't downloaded. Download?"

**Comment by biodranik** [comment]
No. You downloaded the map. The search has found something. But the house is not added to OSM. And user sees a pin on an empty map.

**Comment by zyphlar** [comment]
That already happens for tons of address search results inside OSM: any point which is an address but not a building would show an "empty map" according to this logic, which is fine... especially if the map zoom display was appropriate to show nearby streets for sanity checking.

**Comment by biodranik** [comment]
We are talking about users experience in OM. About users who never heard about OSM.  My point is that they could be disappointed in both scenarios: when the search did not find anything, or when the search finds an empty place on the OM map.

**Comment by zyphlar** [comment]
Sure, but I promise you if I search for "1234 Main St" and get nothing, I will be much more upset than if I search for "1234 Main St" and get a map marker on Main St halfway between 12th and 13th Street. The first is unacceptable in any modern GPS, the second is standard behavior in every GPS.

**Comment by biodranik** [comment]
If the street is mapped, then it's more or less ok. But what if it's not?

Finding addresses missing on the map is good, but it requires additional efforts for our small team. So here I see not only usability questions, but also priority issues. Is this feature more important, for example, than more often maps updates? Or Android Auto support? Or GPX import? Or track recording/sharing?

Everyone who wants to contribute is welcome. Just keep in mind that one of the problems of importing addresses is how to filter out duplicates already present in OSM.

**Comment by zyphlar** [comment]
Generally I think yes it's absolutely critical for gps apps to be able to find the locations they're being used to navigate to. You're right duplication wouldn't be a good idea, maybe osmand's "online search" two-step search would be a good example to follow when the user doesn't find what they want.

**Comment by ruffson** [comment]
I have to agree with @zyphlar. Finding locations in the US is simply not possible. 
Someone tells you which address something is at, you plug it into any* OSM app (either number-first or street name-first) and if you are lucky, it will offer you to go to that street. Which is not really helpful because streets here can be looong.

At this point then the app becomes useless because plugging in addresses is one of the most important use-cases of a maps app. If the reason of not doing it is because the result might be blank vs. there is never any result, the former should be preferred.

I understand that the team is small and that the userbase might be bigger in Europe with better data, but it is quite an important feauture for the US.

*The only one that does a half decent job is MagicEarth.

**Comment by zyphlar** [comment]
Yeah I'm pretty sure MagicEarth includes online search like nominatim or photon by default, too. I just want to patronize FOSS clients.

**Comment by nep2ner** [comment]
If OM provided reliable address-->coordinate lookup, regardless of the surrounding area's OSM map, it would - in my opinion - immediately become the world's the best OSM navigation app.

**Comment by DavidKarlas** [comment]
Just want to add in my 2 cents... openaddresses.io is usually available in more developed countries where government is publishing open data... usually such countries also have strong OSM community so roads are 99% mapped, because that is easy to do as armchair mapper, but addresses are almost impossible to map, even buildings take much more time, let's see Melbourne as example: https://www.openstreetmap.org/#map=16/-37.9415/145.0456 it has roads, even named roads, but no addresses... Adding openaddresses.io would boost usefulness of OM a lot in my opinion, similar story in most of Europe and elsewhere where opnaddresses.io is avaiable.

**Comment by zyphlar** [comment]
@biodranik had concerns earlier about what if search results point to an empty part of the map... one way to fix that is much like OsmAnd's unofficial Open Super Map, integrate at the map file level and display the address nodes just like regular OSM address nodes. They can be deduplicated either by ignoring OA points that have the same house number (and unit?) and/or ignoring similar points within a small radius (maybe OSM has 1234 unit A and OA has 1234A, etc)

**Comment by zyphlar** [comment]
The inclusion of TIGER address interpolation greatly reduces the need for this in the US, however it may still be a good idea for precise location of addresses around the globe.
